parsing/lexer.mll

```diff
@@ -598,6 +601,26 @@ rule token = parse
         CHAR '\n' }
   | "\'" ([^ '\\' '\'' '\010' '\013'] as c) "\'"
       { CHAR c }
+  | "\'" ([^ '\\' '\'' '\010' '\013'] as c) "\'"
+      { CHAR c }
+  | "\'" (utf8 as s) "\'"
+      {
+      let d = String.get_utf_8_uchar s 0 in
+      let l = Uchar.utf_decode_length d in
+      if l < String.length s || not (Uchar.utf_decode_is_valid d) then
+        error lexbuf (Invalid_literal s)
+      else
+        let u = Uchar.utf_decode_uchar d in
+        UCHAR u
+      }
+  | "\'U+" ['0'-'9' 'a'-'f' 'A'-'F']['0'-'9' 'a'-'f' 'A'-'F']
+           ['0'-'9' 'a'-'f' 'A'-'F']['0'-'9' 'a'-'f' 'A'-'F'] "\'"
+      { UCHAR (uchar_for_uchar_escape lexbuf ~first:3 ~last:6) }
+  | "\'U+" ['0'-'9' 'a'-'f' 'A'-'F']['0'-'9' 'a'-'f' 'A'-'F']
+           ['0'-'9' 'a'-'f' 'A'-'F']['0'-'9' 'a'-'f' 'A'-'F']
+           ['0'-'9' 'a'-'f' 'A'-'F']['0'-'9' 'a'-'f' 'A'-'F'] "\'"
+      { UCHAR (uchar_for_uchar_escape lexbuf ~first:3 ~last:8) }
```

Comment:
Probably moot if you refactor this to change the syntax, but just in case there's similar code in the adjusted version of the code, it'd be simpler to have a single branch:

```
| "\'U+" (hex_digit hex_digit hex_digit hex_digit
          (hex_digit hex_digit | "") as hex_number) "\'"
```